server/proxy_protocol.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+
+	"github.com/ngaut/log"
+)
+
+const (
+	proxyProtocolV1MaxHeaderLen = 108
+)
+
+var (
+	errProxyProtocolV1HeaderNotValid = errors.New("PROXY Protocol header not valid")
+)
+
+type proxyProtocolDecoder struct {
+	allowAll    bool
+	allowedNets []*net.IPNet
+}
+
+func newProxyProtocolDecoder(allowedIPs string) (*proxyProtocolDecoder, error) {
+	allowAll := false
+	allowedNets := []*net.IPNet{}
+	if allowedIPs == "*" {
+		allowAll = true
+	} else {
+		for _, aip := range strings.Split(allowedIPs, ",") {
+			saip := strings.TrimSpace(aip)
+			_, ipnet, err := net.ParseCIDR(saip)
+			if err == nil {
+				allowedNets = append(allowedNets, ipnet)
+				continue
+			}
+			psaip := fmt.Sprintf("%s/32", saip)
+			_, ipnet, err = net.ParseCIDR(psaip)
+			if err != nil {
+				return nil, err
+			}
+			allowedNets = append(allowedNets, ipnet)
+		}
+	}
+	return &proxyProtocolDecoder{
+		allowAll:    allowAll,
+		allowedNets: allowedNets,
+	}, nil
+}
+
+func (e *proxyProtocolDecoder) checkAllowed(raddr net.Addr) bool {
+	if e.allowAll {
+		return true
+	}
+	taddr, ok := raddr.(*net.TCPAddr)
+	if !ok {
+		return false
+	}
+	cip := taddr.IP
+	for _, ipnet := range e.allowedNets {
+		if ipnet.Contains(cip) {
+			return true
+		}
+	}
+	return false
+}
+
+func (e *proxyProtocolDecoder) getRealClientAddr(conn net.Conn) net.Addr {
```

Comment:
The name of `getxxxxx` is very confusing here. By naming 'get', we usually expect that the interface can be called repeatedly, at any time. However this interface can be called only once per connection, at the beginning of connection.

In addition, "real" is not obvious here. What's the definition of a "real" address? Something like `ClientAddrBehineProxy` is much more clear for code readers.